pkg/httpd/server.go

```diff
@@ -0,0 +1,95 @@
+package httpd
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+var ShutdownTimeout = time.Second * 5
+
+type Server struct {
+	addr   string
+	lnAddr string
+	logger *zap.Logger
+	srv    *http.Server
+	done   chan struct{}
+	err    error
+}
+
+func New(addr string, h http.Handler) *Server {
+	return &Server{addr: addr, srv: &http.Server{Handler: h}, logger: zap.NewNop()}
+}
+
+func (s *Server) SetLogger(l *zap.Logger) {
+	s.logger = l
+}
+
+func (s *Server) Addr() string {
+	return s.lnAddr
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	s.done = make(chan struct{})
+	s.srv.BaseContext = func(l net.Listener) context.Context { return ctx }
+	ln, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return err
+	}
+	s.lnAddr = ln.Addr().String()
+	s.logger.Info("Listening", zap.String("addr", s.lnAddr))
+	go s.serve(ctx, ln)
+	return nil
+}
+
+func (s *Server) Run(ctx context.Context) error {
+	if err := s.Start(ctx); err != nil {
+		return err
+	}
+	return s.Wait()
+}
+
+func (s *Server) Wait() error {
+	if s.done == nil {
+		return errors.New("Start() must be called before Wait() can be called")
+	}
+	<-s.done
+	return s.err
+}
+
+func (s *Server) serve(ctx context.Context, ln net.Listener) {
+	defer close(s.done)
+	errCh := make(chan error)
+	go func() {
+		if err := s.srv.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errCh <- err
+		}
+		close(errCh)
+	}()
+	var srvErr error
+	select {
+	case srvErr = <-errCh:
+	case <-ctx.Done():
+	}
+	timeout, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
```

Comment:
Can you add a log here (and a blank line too) to identify the shift from  'waiting & serving' phase from the 'shutting down'  phase?

Reply:
just wanted to bump the above comment

Reply:
done.

Reply:
@mattnibs : I don't see the change?

Reply:
For some reason it's not rendering here. If you look at the files directly the change is there: https://github.com/brimsec/zq/pull/500/files#diff-a2c896ceeab22af556c3fb6d1c953cc1R77

Reply:
Now I understand. I added a log line for "Shutting down" along with the reason for shutdown.